iox_query_influxql/src/window/percent_row_number.rs

```diff
@@ -0,0 +1,139 @@
+    fn new() -> Self {
+        Self {
+            percentile: None,
+            len: 0,
+        }
+    }
+
+    fn set_percentile(&mut self, array: ArrayRef) -> Result<()> {
+        if self.percentile.is_none() {
+            self.percentile = match array.data_type() {
+                DataType::Int64 => Some(downcast_value!(array, Int64Array).value(0) as f64),
+                DataType::Float64 => Some(downcast_value!(array, Float64Array).value(0)),
+                dt => {
+                    return error::internal(format!(
+                        "invalid data type ({dt}) for PERCENTILE n argument"
+                    ))
+                }
+            };
+        }
+        Ok(())
+    }
+}
+
+impl Accumulator for PercentRowNumberAccumulator {
+    fn update_batch(&mut self, values: &[ArrayRef]) -> Result<()> {
+        assert_eq!(values.len(), 2);
+
+        self.set_percentile(Arc::clone(&values[1]))?;
+        let array = Arc::clone(&values[0]);
+        self.len += array.len() - array.null_count();
+        Ok(())
+    }
+
+    fn evaluate(&self) -> Result<ScalarValue> {
+        Ok(ScalarValue::UInt64(self.percentile.and_then(|n| {
+            percentile_idx(self.len, n).map(|n| n as u64)
+        })))
+    }
+
+    fn size(&self) -> usize {
+        std::mem::size_of::<Option<f64>>() + std::mem::size_of::<usize>()
+    }
+
+    /// `state` is only called when used as an aggregate function. It can be
+    /// can safely left unimplemented, as this accumulator is only used as a window aggregate.
+    ///
+    /// See: <https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.Accumulator.html#tymethod.state>
+    fn state(&self) -> Result<Vec<ScalarValue>> {
+        error::internal("unexpected call to DerivativeAccumulator::state")
+    }
+
+    /// `merge_batch` is only called when used as an aggregate function. It can be
+    /// can safely left unimplemented, as this accumulator is only used as a window aggregate.
+    ///
+    /// See: <https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.Accumulator.html#tymethod.state>
+    fn merge_batch(&mut self, _states: &[ArrayRef]) -> Result<()> {
+        error::internal("unexpected call to DerivativeAccumulator::merge_batch")
+    }
+}
+
+/// Calculate the location in an ordered list of len items where the
+/// location of the item at the given percentile would be found.
+///
+/// Note that row numbers are 1-based so this returns values in the
+/// range \[1,len\].
+///
+/// This uses the same algorithm as the original influxdb implementation
+/// of percentile as can be found in
+/// <https://github.com/influxdata/influxdb/blob/75a8bcfae2af7b0043933be9f96b98c0741ceee3/influxql/query/call_iterator.go#L1087>.
+fn percentile_idx(len: usize, percentile: f64) -> Option<usize> {
+    match TryInto::<usize>::try_into(((len as f64) * percentile / 100.0 + 0.5).floor() as isize) {
+        Ok(idx) if 0 < idx && idx < len => Some(idx),
+        _ => None,
+    }
+}
```

Comment:
Given it isn't possible for the expression:

```rust
((len as f64) * percentile / 100.0 + 0.5).floor() as isize
```

to be negative, `try_into` will always succeed, so you could simplify it a bit:

```suggestion
fn percentile_idx(len: usize, percentile: f64) -> Option<usize> {
    match ((len as f64) * percentile / 100.0 + 0.5).floor() as usize {
        idx if 0 < idx && idx < len => Some(idx),
        _ => None,
    }
}
```

Based on my interpretation of the InfluxQL OG implementation, it seems that for a small number of rows, low percentiles will return an empty result. For example, say `len` was `10` and you wanted to return the `1st` percentile, this function would return `None`. Equally, the InfluxQL OG would return a `nil`. Does that seem right to you?

Reply:
> Given it isn't possible for the expression:
> 
> ((len as f64) * percentile / 100.0 + 0.5).floor() as isize
>
> to be negative, try_into will always succeed, so you could simplify it a bit:

It is possible if the `n` argument is negative. Whilst this is a meaningless thing to do, as far as I can tell there is nothing that stops it.

Reply:
> Based on my interpretation of the InfluxQL OG implementation, it seems that for a small number of rows, low percentiles will return an empty result. For example, say len was 10 and you wanted to return the 1st percentile, this function would return None. Equally, the InfluxQL OG would return a nil. Does that seem right to you?

In my experiments `PERCENTILE(..., 0)` always returns an empty result. My analysis is the same, low enough percentiles don't yield results.I'm not certain about this being "right", but it is consistent.

Reply:
> It is possible if the n argument is negative. Whilst this is a meaningless thing to do, as far as I can tell there is nothing that stops it.

Ugh – right, I thought it was validated in the `check_percentile` function, but alas it is not, nor is the original. We could probably do that in the future, but certainly for compatibility we can leave as-is.

Reply:
> In my experiments `PERCENTILE(..., 0)` always returns an empty result. My analysis is the same, low enough percentiles don't yield results.I'm not certain about this being "right", but it is consistent.

That was my previous experience too 👍🏻